File: sort012.cpp
#include <iostream>
using namespace std;

void sortArray(int arr[], int n){
    int low = 0;
    int mid = 0;
    int high  = n-1;
    while(mid<=high){
        if(arr[mid] == 0){
            swap(arr[mid],arr[low]);
            mid++;
            low++;
        }
        else if(arr[mid] == 1){
            mid++;
        }
        else {
            swap(arr[mid],arr[high]);
            high--;
        }
    }
}


void printArray(int arr[], int n){
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}


int main(){

    int arr[10] = {0,0,1,0,1,0,1,2,0,1};
    int arr2[10] = {0,1,2,0,1,2,0,1,2,0};

    sortArray(arr2,10);
    printArray(arr2,10);
    

    return 0;
}

File: C++/array/JAVA/Sort_0_1_2.java
class Sort_0_1_2
{
    public static void main(String args[])
    {
        int arr[] = {1,2,0,2,1,2,0,1};
        int n = arr.length;
        int l=0, mid = 0, high = n-1, temp = 0;
        while(mid <=high)
        {
            switch(arr[mid]){
                case 0: temp = arr[l];
                        arr[l] = arr[mid];
                        arr[mid] = temp;
                        mid++;
                        l++;
                        break;
                case 1: mid++;
                        break;
                case 2: temp = arr[high];
                        arr[high] = arr[mid];
                        arr[mid] = temp;
                        high--;
                        break;
            }
        }
        for(int i=0;i<n;i++)
        {
            System.out.print(arr[i] + " ");
        }
    }
}